m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,781 @@
+
+                    tags.put(bucketIdTagName, idTagValue);
+                    tags.put(bucketTagName,
+                        String.format("%s-%s",
+                            durationBucketString(bucketPairs[i].lowerBoundDuration()),
+                            durationBucketString(bucketPairs[i].upperBoundDuration())
+                        )
+                    );
+
+                    break;
+                }
+            }
+
+            reportCounterInternal(name, tags, samples);
+        }
+    }
+
+    // Relies on the calling function to provide guarantees of the reporter being open
+    private void reportCounterInternal(String name, Map<String, String> tags, long value) {
+        CountValue countValue = new CountValue();
+        countValue.setI64Value(value);
+
+        MetricValue metricValue = new MetricValue();
+        metricValue.setCount(countValue);
+
+        Metric metric = newMetric(name, tags, metricValue);
+
+        queueSizedMetric(new SizedMetric(metric, calculateSize(metric)));
+    }
+
+    private Metric newMetric(String name, Map<String, String> tags, MetricValue metricValue) {
+        Metric metric = new Metric(name);
+        metric.setTags(toMetricTagSet(tags));
+        metric.setTimestamp(System.currentTimeMillis() * Duration.NANOS_PER_MILLI);
+        metric.setMetricValue(metricValue);
+
+        return metric;
+    }
+
+    private void queueSizedMetric(SizedMetric sizedMetric) {
+        try {
+            metricQueue.put(sizedMetric);
+        } catch (InterruptedException e) {
+            LOG.warn(String.format("Interrupted queueing metric: %s", sizedMetric.getMetric().getName()));
+        }
+    }
+
+    private class Processor implements Runnable {
+        List<Metric> pendingMetrics = new ArrayList<>(freeBytes / 10);
+        int metricsSize = 0;
+
+        @Override
+        public void run() {
+            try {
+                while (!executor.isShutdown()) {
+                    // This `take` call will block until there is an item on the queue to take.
+                    // When this reporter is closed, shutdownNow will be called on the executor,
+                    // which will interrupt this thread and proceed to the `InterruptedException`
+                    // catch block.
+                    SizedMetric sizedMetric = metricQueue.poll(MAX_PROCESSOR_WAIT_UNTIL_FLUSH_MILLIS, TimeUnit.MILLISECONDS);
+
+                    if (sizedMetric != null) {
+                        flushMetricIteration(sizedMetric);
+                    } else if (!pendingMetrics.isEmpty()) {
+                        // If we don't get any new metrics after waiting the specified time,
+                        // flush what we have so far.
+                        flushMetricIteration(SizedMetric.FLUSH);
+                    }
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
```

Comment:
Sorry I'm missing something. A processor is already running on a different thread, and it's received an interrupt, why does it need to interrupt itself again?

Reply:
Oops good catch - missed this earlier.